### ISSUE TITLE
Add IReactorWin32Event

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4523]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4523|
|Type|enhancement|
|Created|2010-06-29 14:41:07Z|
|Branch|https://github.com/twisted/twisted/tree/ireactorwin32-4523|

`win32eventreactor` has `addEvent` and `removeEvent` methods which are generally useful for Windows-specific stuff, plus useful in the Windows implementation of cross-platform APIs such as `twisted.internet.serialport`.  It would be great if these methods were part of a documented, declared interface (eg, [#3802](https://github.com/twisted/twisted/issues/3802) wants to inspect the reactor for this to provide better failure messages when serialport is used with the wrong reactor).

<details><summary>Searchable metadata</summary>

```
trac-id__4523 4523
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_ireactorwin32_4523 branches/ireactorwin32-4523
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1277822467000000 1277822467000000
changetime__1277859560000000 1277859560000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [29427]) Branching to 'ireactorwin32-4523'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

I pulled the implementation of this out of the [#3802](https://github.com/twisted/twisted/issues/3802) branch.  Very short and simple, please review.  [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ireactorwin32-4523)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli set owner to @exarkun|
|-|-|

Two minor points that may or may not be valid:
 1. The last half of the news fragment is awkwardly phrased ("and twisted.internet.win32eventreactor declares that it implements it.") - the two its are confusing, since each one refers to a different antecedent.  Perhaps change the last "it" to "IReactorWin32Events", or change the whole thing "twisted.internet.interfaces.IReactorWin32Events is now explicitly defined and is declared to be implemented by twisted.internet.win32eventreactor" (although passive voice may not be better).
 1. Should test_win32events be a declared test case for win32eventreactor?  (Should "-*- test-case-name: twisted.internet.test.test_win32events -*-" be at the top of the file?)

Other than that, looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

  1. Rephrased the news fragment to avoid the conjunction entirely
  1. Since tons of other tests still apply, I don't think there's a reason to have a test-case-name on win32eventreactor.py

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [29477]) Merge ireactorwin32-4523

Author: ltaylor.volks, exarkun
Reviewer: cyli
Fixes: [#4523](https://github.com/twisted/twisted/issues/4523)

Introduce a new interface for the Windows event support win32eventreactor
already had.  Also add a simple test for the existing implementation and
make the reactor declare that the interface is indeed implemented.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
